tests/mptcpwrap-tester.c

```diff
@@ -14,34 +14,77 @@
 #include <sys/socket.h>
 #include <netinet/in.h>
 #include <unistd.h>
+#include <stdbool.h>
 
-int main()
-{
-        /*
-          libmptcpwrap.so should be preloaded when running this
-          program, e.g.:
 
-              LD_PRELOAD=libmptcpwrap.so ./mptcpwrap-tester
-        */
-        int const fd = socket(AF_INET, SOCK_STREAM, 0);
-        assert(fd != -1);
+struct socket_data
+{
+        int  domain;
+        int  type;
+        int  protocol;
+        bool expect_mptcp;
+};
 
+static bool verify_protocol(int fd, bool expect_mptcp)
+{
         int protocol = 0;
         socklen_t len = sizeof(protocol);
 
-        int const expected_protocol = IPPROTO_TCP + 256;  // MPTCP-ized.
-
         int const ret = getsockopt(fd,
                                    SOL_SOCKET,
                                    SO_PROTOCOL,
                                    &protocol,
                                    &len);
 
+        /*
+          The MPTCP-ized protocol injected by the hijacked socket()
+          call, i.e. IPPROTO_MPTCP.
+        */
+        static int const mptcp_protocol = IPPROTO_TCP + 256;
+
+        return ret == 0
+                && (expect_mptcp ? protocol == mptcp_protocol : true);
+}
+
+static void test_socket_data(struct socket_data const *data)
+{
+        /*
+          libmptcpwrap.so should be preloaded when running this
+          program, e.g.:
+
+          LD_PRELOAD=libmptcpwrap.so ./mptcpwrap-tester
+        */
+        int const fd = socket(data->domain, data->type, data->protocol);
+        assert(fd != -1);
+
+        bool const verified = verify_protocol(fd, data->expect_mptcp);
+
         close(fd);
 
-        assert(ret == 0
-               && protocol == expected_protocol
-               && len == sizeof(protocol));
+        assert(verified);
+}
+
+int main()
+{
+        /*
+          MPTCP is only injected when using the SOCK_STREAM socket
+          type and a protocol value of 0 or IPPROTO_TCP.
+        */
+        static struct socket_data const data[] = {
+                { AF_LOCAL, SOCK_STREAM, 0,            false },
+                { AF_INET,  SOCK_STREAM, IPPROTO_SCTP, false },
+                { AF_INET,  SOCK_DGRAM,  0,            false },
+                { AF_INET,  SOCK_STREAM, 0,            true  },
+                { AF_INET6, SOCK_STREAM, 0,            true  },
+                { AF_INET,  SOCK_STREAM, IPPROTO_TCP,  true  },
+                { AF_INET6, SOCK_STREAM, IPPROTO_TCP,  true  }
+        };
+
+        static struct socket_data const *const end =
+                data + (sizeof(data) / sizeof(data[0]));
+
+        for (struct socket_data const *d = data; d != end; ++d)
+                test_socket_data(d);
```

Comment:
When running tests in a loop like this, the failure output looks the same for every case so it takes some work to figure out what failed. Something simple like printing the array index on failure will help!

Reply:
Done.